## The Battle of the Neighborhoods - Week 5 - Boroughs of New York City

### Segmenting and Clustering Neighborhoods -  Brooklyn and Manhattan

## Introduction

In this section of the capstone project, we will use the Foursquare API to explore neighborhoods in Brooklyn and Manhattan. We will use the **explore** function to get the most common venue categories in each neighborhood, and then use this feature to group the neighborhoods into clusters. We will use the *k*-means clustering algorithm to complete this task. Finally, we will use the Folium library to visualize the neighborhoods in Brooklyn and Manhattan and their emerging clusters.

## Table of Contents<Br>

***1. Download and Explore Dataset***<Br>
***2. Explore Neighborhoods in Brooklyn and Manhattan***<Br>
***3. Analyze Each Neighborhood***<Br>
***4. Cluster Neighborhoods and Examine Clusters***<Br>

***Download all the dependencies that are needed.***

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

#!conda install -c conda-forge geopy --yes
# uncomment this line if you haven't completed the Foursquare API lab
from geopy.geocoders import Nominatim 
# convert an address into latitude and longitude values

import requests 
# library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors
import matplotlib.pyplot as plt

# import k-means from clustering stage
from sklearn.cluster import KMeans

from sklearn.metrics import silhouette_score

#!conda install -c conda-forge 
#folium=0.5.0 --yes 
# uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 1. Download and Explore Dataset

Neighborhood has a total of 5 boroughs and 306 neighborhoods. In order to segement the neighborhoods and explore them, we will essentially need a dataset that contains the 5 boroughs and the neighborhoods that exist in each borough as well as the latitude and logitude coordinates of each neighborhood.

***Load and explore the data***

In [4]:
NYC_Geo=pd.read_csv('BON1_NYC_GEO.csv')
print('Data downloaded!')

Data downloaded!


In [ ]:
NYC_Geo.head()

In [ ]:
NYC_Geo['Borough'].value_counts().to_frame()

In [7]:
NYC_Geo.shape

(306, 4)

In [8]:
print(NYC_Geo.Borough.unique())

['Bronx' 'Manhattan' 'Brooklyn' 'Queens' 'Staten Island']


In [ ]:
NYC_Geo.isnull().sum()

### Segmenting and Clustering Neighborhoods -  Brooklyn and Manhattan

In [10]:
BM_Geo = NYC_Geo.loc[(NYC_Geo['Borough'] == 'Brooklyn')|(NYC_Geo['Borough'] == 'Manhattan')]
BM_Geo = BM_Geo.reset_index(drop=True)
BM_Geo.head()

Borough Neighborhood   Latitude  Longitude
0  Manhattan  Marble Hill  40.876551 -73.910660
1   Brooklyn    Bay Ridge  40.625801 -74.030621
2   Brooklyn  Bensonhurst  40.611009 -73.995180
3   Brooklyn  Sunset Park  40.645103 -74.010316
4   Brooklyn   Greenpoint  40.730201 -73.954241

In [11]:
BM_Geo.shape

(110, 4)

***Use geopy library to get the latitude and longitude values of New York City. The geograpical coordinate of New York City are 40.7308619, -73.9871558..***

In [12]:
import time
start_time = time.time()

address = 'New York City, NY'

geolocator = Nominatim(user_agent="Jupyter")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of New York City are {}, {}.'.format(latitude, longitude))

print("--- %s seconds ---" % round((time.time() - start_time), 2))

The geograpical coordinate of New York City are 40.7127281, -74.0060152.
--- 1.55 seconds ---


***Create a map of Brooklyn  and Manhattan with neighborhoods superimposed on top.***

In [ ]:
# create map of Toronto using latitude and longitude values
map_BM = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(BM_Geo['Latitude'], BM_Geo['Longitude'], BM_Geo['Borough'], BM_Geo['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_BM)  
    
map_BM

***Define Foursquare Credentials and Version***

In [14]:
CLIENT_ID='4WOAAQII3QSIJH1NQLJAKLR3CVU2CXZQAJI215NWI3223KXN'# your Foursquare I
CLIENT_SECRET='BJQS0NNC5Y3IG2NTRFTCKFGMDYSQPEILPHIG03Y0SMTZL0FW'# your Foursquare 
VERSION = '20181218' # Foursquare API version

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 4WOAAQII3QSIJH1NQLJAKLR3CVU2CXZQAJI215NWI3223KXN
CLIENT_SECRET:BJQS0NNC5Y3IG2NTRFTCKFGMDYSQPEILPHIG03Y0SMTZL0FW


## 2. Explore Neighborhoods in Brooklyn and Manhattan

#### Extract Venues data for each neighborhoods in Brooklyn

In [15]:
def getNearbyVenues(names, latitudes, longitudes, LIMIT=200, radius=1000):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

***Run the above function on each neighborhood and create a new dataframe called BM_venues.***

In [ ]:
BM_venues = getNearbyVenues(names=BM_Geo['Neighborhood'],
                                  latitudes=BM_Geo['Latitude'],
                                  longitudes=BM_Geo['Longitude'],
                                  LIMIT=200)

print('The "BM_venues" dataframe has {} venues and {} unique venue types.'.format(
      len(BM_venues['Venue Category']),
      len(BM_venues['Venue Category'].unique())))

BM_venues.to_csv('BM_venues.csv', sep=',', encoding='UTF8')
BM_venues.head()

***Download data from BM_venues.csv***

In [17]:
colnames = ['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']
BM_venues = pd.read_csv('BM_venues.csv', skiprows=1, names=colnames)
BM_venues.columns = BM_venues.columns.str.replace(' ', '')
BM_venues.head()

Neighborhood  NeighborhoodLatitude  NeighborhoodLongitude          Venue  \
0  Marble Hill             40.876551              -73.91066    Bikram Yoga   
1  Marble Hill             40.876551              -73.91066       Arturo's   
2  Marble Hill             40.876551              -73.91066  Tibbett Diner   
3  Marble Hill             40.876551              -73.91066    Sam's Pizza   
4  Marble Hill             40.876551              -73.91066      Starbucks   

   VenueLatitude  VenueLongitude VenueCategory  
0      40.876844      -73.906204   Yoga Studio  
1      40.874412      -73.910271   Pizza Place  
2      40.880404      -73.908937         Diner  
3      40.879435      -73.905859   Pizza Place  
4      40.877531      -73.905582   Coffee Shop

In [18]:
BM_venues.shape

(9748, 7)

***Visualize the BM_Venues data***

In [19]:
def Venues_Map(Borough_name, Borough_neighborhoods):
    
    # Use geopy library to get the latitude and longitude values 
    geolocator = Nominatim(user_agent="Jupyter")
    Borough_location = geolocator.geocode(Borough_name) #'Brooklyn, NY'
    Borough_latitude = Borough_location.latitude
    Borough_longitude = Borough_location.longitude
    print('The geographical coordinates of "{}" are {}, {}.'.format(Borough_name, Borough_latitude, Borough_longitude))
    
    # To verify the number of Boroughs and Neighborhoods in the extracted data
    print('The "{}" dataframe has {} different venue types and {} neighborhoods.'.format(
          Borough_name,
          len(Borough_neighborhoods['VenueCategory'].unique()),
          len(Borough_neighborhoods['Neighborhood'].unique())))
    
    # create map of city using latitude and longitude values
    map_Borough = folium.Map(location=[Borough_latitude, Borough_longitude], zoom_start=10)

    # add markers to map
    for lat, lng, venue, category in zip(Borough_neighborhoods['VenueLatitude'], Borough_neighborhoods['VenueLongitude'], Borough_neighborhoods['Venue'], Borough_neighborhoods['VenueCategory']):
        label = '{}, {}'.format(category, venue)
        label = folium.Popup(label, parse_html=True)
        folium.CircleMarker(
            [lat, lng],
            radius=0.1,
            popup=label,
            color='red',
            fill=True,
            fill_color='#FF0000',
            fill_opacity=0.3).add_to(map_Borough)  

    return map_Borough

In [ ]:
Venues_Map('New York City, NY', BM_venues)

In [ ]:
BM_venues.groupby('VenueCategory')['Venue'].count().sort_values(ascending=False)

***Check how many venues were returned for each neighborhood***

In [ ]:
BM_venues.groupby('Neighborhood').count()

***Find out how many unique categories can be curated from all the returned venues***

In [23]:
print('There are {} uniques categories.'.format(len(BM_venues['VenueCategory'].unique())))

There are 408 uniques categories.


## 3. Analyze Each Neighborhood

In [24]:
# one hot encoding
BM_onehot = pd.get_dummies(BM_venues[['VenueCategory']], prefix="", prefix_sep="")

#column lists before adding neighborhood
column_names = ['Neighborhood'] + list(BM_onehot.columns)

# add neighborhood column back to dataframe
BM_onehot['Neighborhood'] = BM_venues['Neighborhood'] 

# move neighborhood column to the first column
BM_onehot = BM_onehot[column_names]

BM_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  African Restaurant  \
0  Marble Hill                  0               0                   0   
1  Marble Hill                  0               0                   0   
2  Marble Hill                  0               0                   0   
3  Marble Hill                  0               0                   0   
4  Marble Hill                  0               0                   0   

   Airport Terminal  American Restaurant  Animal Shelter  Antique Shop  \
0                 0                    0               0             0   
1                 0                    0               0             0   
2                 0                    0               0             0   
3                 0                    0               0             0   
4                 0                    0               0             0   

   Aquarium  Arcade  Arepa Restaurant  Argentinian Restaurant  Art Gallery  \
0         0       0                 0                       0            0   
1         0       0                 0                       0            0   
2         0       0                 0                       0            0   
3         0       0                 0                       0            0   
4         0       0                 0                       0            0   

   Art Museum  Arts & Crafts Store  Arts & Entertainment  Asian Restaurant  \
0           0                    0                     0                 0   
1           0                    0                     0                 0   
2           0                    0                     0                 0   
3           0                    0                     0                 0   
4           0                    0                     0                 0   

   Athletics & Sports  Auditorium  Australian Restaurant  Austrian Restaurant  \
0                   0           0                      0                    0   
1                   0           0                      0                    0   
2                   0           0                      0                    0   
3                   0           0                      0                    0   
4                   0           0                      0                    0   

   Auto Workshop  Automotive Shop  BBQ Joint  Bagel Shop  Bakery  Bank  Bar  \
0              0                0          0           0       0     0    0   
1              0                0          0           0       0     0    0   
2              0                0          0           0       0     0    0   
3              0                0          0           0       0     0    0   
4              0                0          0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Bath House  Beach  Bed & Breakfast  Beer Bar  Beer Garden  Beer Store  \
0           0      0                0         0            0           0   
1           0      0                0         0            0           0   
2           0      0                0         0            0           0   
3           0      0                0         0            0           0   
4           0      0                0         0            0           0   

   Big Box Store  Bike Rental / Bike Share  Bike Shop  Bike Trail  Bistro  \
0              0                         0          0           0       0   
1              0                         0          0           0       0   
2              0                         0          0           0  

***To filter only Restaurant data***

In [25]:
restaurant_List = []
search = 'Restaurant'
for i in BM_onehot.columns :
    if search in i:
        restaurant_List.append(i)

In [ ]:
restaurant_List

In [27]:
col_name = []
col_name = ['Neighborhood'] + restaurant_List
BM_restaurant = BM_onehot[col_name]
BM_restaurant = BM_restaurant.iloc[:,1::]

In [28]:
BM_restaurant_grouped = BM_restaurant.groupby('Neighborhood').sum().reset_index()

In [29]:
BM_restaurant_grouped['Total'] = BM_restaurant_grouped .sum(axis=1)

## 4. Cluster Neighborhoods and Examine Clusters

***First, let's determine the optimal value of K for our dataset using the Silhouette Coefficient Method***

***From sklearn documentation - https://scikit-learn.org/stable/modules/clustering.html#silhouette-coefficient***

A higher Silhouette Coefficient score relates to a model with better-defined clusters. The Silhouette Coefficient is defined for each sample and is composed of two scores: `

    a: The mean distance between a sample and all other points in the same class.

    b: The mean distance between a sample and all other points in the next nearest cluster.

The Silhouette Coefficient is for a single sample is then given as:

s=b-a/max(a,b)

Now, to find the optimal value of k for KMeans, loop through 1..n for n_clusters in KMeans and calculate Silhouette Coefficient for each sample.

A higher Silhouette Coefficient indicates that the object is well matched to its own cluster and poorly matched to neighboring clusters.

In [30]:
BM_grouped_clustering = BM_restaurant_grouped.drop('Neighborhood', 1)

for n_cluster in range(2, 10):
    kmeans = KMeans(n_clusters=n_cluster).fit(BM_grouped_clustering)
    label = kmeans.labels_
    sil_coeff = silhouette_score(BM_grouped_clustering, label, metric='euclidean')
    print("For n_clusters={}, The Silhouette Coefficient is {}".format(n_cluster, sil_coeff))

For n_clusters=2, The Silhouette Coefficient is 0.368742986483858
For n_clusters=3, The Silhouette Coefficient is 0.3015724167964779
For n_clusters=4, The Silhouette Coefficient is 0.21680256528839
For n_clusters=5, The Silhouette Coefficient is 0.23282933915106716
For n_clusters=6, The Silhouette Coefficient is 0.23621402475430503
For n_clusters=7, The Silhouette Coefficient is 0.1935325133895177
For n_clusters=8, The Silhouette Coefficient is 0.18935337809341524
For n_clusters=9, The Silhouette Coefficient is 0.1778681930501573


***As we can see, n_clusters=2 has highest Silhouette Coefficient. This means that 2 should be the optimal number of clusters.<br>
For n_clusters=2, The Silhouette Coefficient is 0.368742986483858***

***Run k-means to cluster the neighborhood into 2 clusters.***

In [31]:
# set number of clusters
kclusters = 2

BM_grouped_clustering = BM_restaurant_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(BM_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 1, 0, 1, 0,
       1, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0,
       0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1,
       0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 1])

In [32]:
BM_results = pd.DataFrame(kmeans.cluster_centers_)
BM_results.columns = BM_grouped_clustering.columns
BM_results.index = ['cluster0','cluster1']
BM_results['Total Sum'] = BM_results.sum(axis = 1)
BM_results

African Restaurant  American Restaurant  Arepa Restaurant  \
cluster0            0.037037             1.166667          0.018519   
cluster1            0.089286             2.071429          0.125000   

          Argentinian Restaurant  Asian Restaurant  Australian Restaurant  \
cluster0                0.018519          0.185185          -9.714451e-17   
cluster1                0.125000          0.642857           1.607143e-01   

          Austrian Restaurant  Brazilian Restaurant  Burmese Restaurant  \
cluster0             0.055556         -1.734723e-18       -1.734723e-18   
cluster1             0.071429          1.785714e-02        1.785714e-02   

          Cajun / Creole Restaurant  Cambodian Restaurant  \
cluster0                   0.037037         -6.938894e-18   
cluster1                   0.053571          5.357143e-02   

          Cantonese Restaurant  Caribbean Restaurant  Caucasian Restaurant  \
cluster0          4.163336e-17              1.574074              0.055556   
cluster1          1.250000e-01              1.089286              0.053571   

          Chinese Restaurant  Colombian Restaurant  Comfort Food Restaurant  \
cluster0            1.203704         -1.734723e-18                 0.037037   
cluster1            1.339286          1.785714e-02                 0.071429   

          Cuban Restaurant  Czech Restaurant  Dumpling Restaurant  \
cluster0          0.148148     -3.469447e-18        -1.387779e-17   
cluster1          0.250000      3.571429e-02         1.428571e-01   

          Eastern European Restaurant  Empanada Restaurant  \
cluster0                     0.055556        -6.938894e-18   
cluster1                     0.125000         7.142857e-02   

          English Restaurant  Ethiopian Restaurant  Falafel Restaurant  \
cluster0       -3.469447e-18              0.018519            0.111111   
cluster1        3.571429e-02              0.196429            0.196429   

          Fast Food Restaurant  Filipino Restaurant  French Restaurant  \
cluster0              1.037037             0.018519           0.351852   
cluster1              0.250000             0.125000           1.142857   

          German Restaurant  Greek Restaurant  Halal Restaurant  \
cluster0           0.055556          0.055556          0.037037   
cluster1           0.107143          0.553571          0.035714   

          Hawaiian Restaurant  Hotpot Restaurant  Indian Restaurant  \
cluster0         2.775558e-17      -3.469447e-18           0.203704   
cluster1         8.928571e-02       3.571429e-02           1.035714   

          Israeli Restaurant  Italian Restaurant  Japanese Curry Restaurant  \
cluster0        2.775558e-17            1.222222                   0.055556   
cluster1        8.928571e-02            3.214286                   0.071429   

          Japanese Restaurant  Jewish Restaurant  Kebab Restaurant  \
cluster0             0.685185      -6.938894e-18          0.018519   
cluster1             1.517857       5.357143e-02          0.017857   

          Korean Restaurant  Kosher Restaurant  Latin American Restaurant  \
cluster0           0.018519           0.018519                   0.407407   
cluster1           0.482143           0.035714                   0.892857   

          Lebanese Restaurant  Malay Restaurant  Mediterranean Restaurant  \
cluster0             0.037037     -6.938894e-18                  0.277778   
cluster1             0.071429      7.142857e-02                  0.696429   

          Mexican Restaurant  Middle Eastern Restaurant  \
cluster0            0.685185                   0.296296   
cluster1            2.000000                   0.464286   

          Modern European Restaurant  Molecular Gastronomy Restaurant  \
cluster0               -1.734723e-18                    -3.469447e-18   
cluster1                1.785714e-02                     3.571429e-02   

          Moroccan Restaurant  New American Restaurant  Paella Restaurant  \
cluster0        -6.938894e-18       

The Total and Total Sum of cluster 1 has smallest value. It shows that the market is not saturated.

***Dataframe with Neighborhood,Cluster No and Total Sum***

In [33]:
BM_results_merged = pd.DataFrame(BM_restaurant_grouped['Neighborhood'])

BM_results_merged['Total'] = BM_restaurant_grouped['Total']
BM_results_merged = BM_results_merged.assign(Cluster_Labels = kmeans.labels_)

In [ ]:
print(BM_results_merged.shape)
BM_results_merged

***Merge BM_results_merged with BM_Geo***

In [ ]:
BM_merged = BM_Geo

BM_merged = BM_merged.join(BM_results_merged.set_index('Neighborhood'), on='Neighborhood')

print(BM_merged.shape)
BM_merged.head(10) # check the last columns!

***Finally, let's visualize the resulting clusters***

In [ ]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(BM_merged['Latitude'], BM_merged['Longitude'], BM_merged['Neighborhood'], BM_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## List Neighborhoods of Interest in New York City

## Cluster 1 : Saturated Markets

In [ ]:
BM_merged[BM_merged['Cluster_Labels'] == 1].reset_index(drop=True)

## Cluster 0 : Untapped Markets

In [38]:
BM_merged[BM_merged['Total'] == 0].reset_index(drop=True)

Borough Neighborhood   Latitude  Longitude  Total  Cluster_Labels
0  Brooklyn     Sea Gate  40.576375 -74.007873      0               0

***There are no Untapped Markets for Restaurant business in Brooklyn and Manhattan.***

### Part 5B Segmenting and Clustering Neighborhoods -  Bronx, Queens and Staten Island

## Introduction

In this section of the capstone project, we will use the Foursquare API to explore neighborhoods in Bronx, Queens and Staten Island.

## Table of Contents<Br>

***1. Download and Explore Dataset***<Br>
***2. Explore Neighborhoods in  Bronx, Queens and Staten Island***<Br>
***3. Analyze Each Neighborhood***<Br>
***4. Cluster Neighborhoods and Examine Clusters***<Br>

## 1. Download and Explore Dataset

In [40]:
BQS_Geo = NYC_Geo.loc[(NYC_Geo['Borough'] == 'Bronx')|(NYC_Geo['Borough'] == 'Queens')|(NYC_Geo['Borough'] == 'Staten Island')]
BQS_Geo = BQS_Geo.reset_index(drop=True)
BQS_Geo.head()

Borough Neighborhood   Latitude  Longitude
0   Bronx    Wakefield  40.894705 -73.847201
1   Bronx   Co-op City  40.874294 -73.829939
2   Bronx  Eastchester  40.887556 -73.827806
3   Bronx    Fieldston  40.895437 -73.905643
4   Bronx    Riverdale  40.890834 -73.912585

In [41]:
BQS_Geo.shape

(196, 4)

***Create a map of Bronx, Queens and Staten Island with neighborhoods superimposed on top.***

In [ ]:
# create map of Toronto using latitude and longitude values
map_BQS = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(BQS_Geo['Latitude'], BQS_Geo['Longitude'], BQS_Geo['Borough'], BQS_Geo['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_BQS)  
    
map_BQS

## 2. Explore Neighborhoods in  Bronx, Queens and Staten Island¶

***Run the above function on each neighborhood and create a new dataframe called BQS_venues.***

In [ ]:
BQS_venues = getNearbyVenues(names=BQS_Geo['Neighborhood'],
                                  latitudes=BQS_Geo['Latitude'],
                                  longitudes=BQS_Geo['Longitude'],
                                  LIMIT=200)

print('The "BQS_venues" dataframe has {} venues and {} unique venue types.'.format(
      len(BQS_venues['Venue Category']),
      len(BQS_venues['Venue Category'].unique())))

BQS_venues.to_csv('BQS_venues.csv', sep=',', encoding='UTF8')
BQS_venues.head()

***Download data from Brooklyn_venues.csv***

In [47]:
colnames = ['Neighborhood', 'Neighborhood Latitude', 'Neighborhood Longitude', 'Venue', 'Venue Latitude', 'Venue Longitude', 'Venue Category']
BQS_venues = pd.read_csv('BQS_venues.csv', skiprows=1, names=colnames)
BQS_venues.columns = BQS_venues.columns.str.replace(' ', '')
BQS_venues.head()

Neighborhood  NeighborhoodLatitude  NeighborhoodLongitude  \
0    Wakefield             40.894705             -73.847201   
1    Wakefield             40.894705             -73.847201   
2    Wakefield             40.894705             -73.847201   
3    Wakefield             40.894705             -73.847201   
4    Wakefield             40.894705             -73.847201   

                         Venue  VenueLatitude  VenueLongitude  \
0             Lollipops Gelato      40.894123      -73.845892   
1           Ripe Kitchen & Bar      40.898152      -73.838875   
2              Ali's Roti Shop      40.894036      -73.856935   
3                     Rite Aid      40.896649      -73.844846   
4  Jackie's West Indian Bakery      40.889283      -73.843310   

          VenueCategory  
0          Dessert Shop  
1  Caribbean Restaurant  
2  Caribbean Restaurant  
3              Pharmacy  
4  Caribbean Restaurant

***Visualize the BQS_Venues data***

In [ ]:
Venues_Map('New York City, NY', BQS_venues)

In [ ]:
BQS_venues.groupby('VenueCategory')['Venue'].count().sort_values(ascending=False)

***Let's check how many venues were returned for each neighborhood***

In [ ]:
BQS_venues.groupby('Neighborhood').count()

***Let's find out how many unique categories can be curated from all the returned venues***

In [51]:
print('There are {} uniques categories.'.format(len(BQS_venues['VenueCategory'].unique())))

There are 387 uniques categories.


## 3. Analyze Each Neighborhood

In [52]:
# one hot encoding
BQS_onehot = pd.get_dummies(BQS_venues[['VenueCategory']], prefix="", prefix_sep="")

#column lists before adding neighborhood
column_names = ['Neighborhood'] + list(BQS_onehot.columns)

# add neighborhood column back to dataframe
BQS_onehot['Neighborhood'] = BQS_venues['Neighborhood'] 

# move neighborhood column to the first column
BQS_onehot = BQS_onehot[column_names]

BQS_onehot.head()

Neighborhood  ATM  Accessories Store  Afghan Restaurant  African Restaurant  \
0    Wakefield    0                  0                  0                   0   
1    Wakefield    0                  0                  0                   0   
2    Wakefield    0                  0                  0                   0   
3    Wakefield    0                  0                  0                   0   
4    Wakefield    0                  0                  0                   0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Arcade  Arepa Restaurant  Argentinian Restaurant  \
0             0       0                 0                       0   
1             0       0                 0                       0   
2             0       0                 0                       0   
3             0       0                 0                       0   
4             0       0                 0                       0   

   Art Gallery  Art Museum  Arts & Crafts Store  Arts & Entertainment  \
0            0           0                    0                     0   
1            0           0                    0                     0   
2            0           0                    0                     0   
3            0           0                    0                     0   
4            0           0                    0                     0   

   Asian Restaurant  Astrologer  Athletics & Sports  Australian Restaurant  \
0                 0           0                   0                      0   
1                 0           0                   0                      0   
2                 0           0                   0                      0   
3                 0           0                   0                      0   
4                 0           0                   0                      0   

   Auto Dealership  Auto Garage  Auto Workshop  Automotive Shop  BBQ Joint  \
0                0            0              0                0          0   
1                0            0              0                0          0   
2                0            0              0                0          0   
3                0            0              0                0          0   
4                0            0              0                0          0   

   Bagel Shop  Bakery  Bank  Bar  Baseball Field  Baseball Stadium  \
0           0       0     0    0               0                 0   
1           0       0     0    0               0                 0   
2           0       0     0    0               0                 0   
3           0       0     0    0               0                 0   
4           0       0     0    0               0                 0   

   Basketball Court  Bath House  Beach  Beach Bar  Bed & Breakfast  Beer Bar  \
0                 0           0      0          0                0         0   
1                 0           0      0          0                0         0   
2                 0           0      0          0                0         0   
3                 0           0      0          0                0         0   
4                 0           0      0          0                0         0   

   Beer Garden  Beer Store  Big Box Store  Bike Rental / Bike Share  \
0            0           0              0                         0   
1            0           0              0                         0   
2            0           0              0                         0   
3            0           0              0                         0   
4  

In [53]:
restaurant_List1 = []
search = 'Restaurant'
for i in BQS_onehot.columns :
    if search in i:
        restaurant_List1.append(i)

In [ ]:
restaurant_List1

In [55]:
col_name = []
col_name = ['Neighborhood'] + restaurant_List1
BQS_restaurant = BQS_onehot[col_name]
BQS_restaurant = BQS_restaurant.iloc[:,1::]

In [56]:
BQS_restaurant_grouped = BQS_restaurant.groupby('Neighborhood').sum().reset_index()

In [57]:
BQS_restaurant_grouped['Total'] = BQS_restaurant_grouped .sum(axis=1)

## 4. Cluster Neighborhoods and Examine Clusters

***First, let's determine the optimal value of K for our dataset using the Silhouette Coefficient Method***

In [58]:
BQS_grouped_clustering = BQS_restaurant_grouped.drop('Neighborhood', 1)

for n_cluster in range(2, 10):
    kmeans = KMeans(n_clusters=n_cluster).fit(BQS_grouped_clustering)
    label = kmeans.labels_
    sil_coeff = silhouette_score(BQS_grouped_clustering, label, metric='euclidean')
    print("For n_clusters={}, The Silhouette Coefficient is {}".format(n_cluster, sil_coeff))

For n_clusters=2, The Silhouette Coefficient is 0.507905521359579
For n_clusters=3, The Silhouette Coefficient is 0.4156824534946386
For n_clusters=4, The Silhouette Coefficient is 0.3457060913809446
For n_clusters=5, The Silhouette Coefficient is 0.3500783270223197
For n_clusters=6, The Silhouette Coefficient is 0.34923358325639453
For n_clusters=7, The Silhouette Coefficient is 0.32273710251642734
For n_clusters=8, The Silhouette Coefficient is 0.280830807763326
For n_clusters=9, The Silhouette Coefficient is 0.28290210804904803


***As we can see, n_clusters=2 has highest Silhouette Coefficient. This means that 2 should be the optimal number of clusters.<br>
For n_clusters=2, The Silhouette Coefficient is 0.507905521359579***

***Run k-means to cluster the neighborhood into 2 clusters.***

In [59]:
# set number of clusters
kclusters = 2

BQS_grouped_clustering = BQS_restaurant_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(BQS_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 1, 0,
       1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0,
       0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 0, 1,
       0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1])

In [60]:
BQS_results = pd.DataFrame(kmeans.cluster_centers_)
BQS_results.columns = BQS_grouped_clustering.columns
BQS_results.index = ['cluster0','cluster1']
BQS_results['Total Sum'] = BQS_results.sum(axis = 1)
BQS_results

Afghan Restaurant  African Restaurant  American Restaurant  \
cluster0           0.006993            0.020979             0.482517   
cluster1           0.039216            0.039216             1.274510   

          Arepa Restaurant  Argentinian Restaurant  Asian Restaurant  \
cluster0          0.013986            1.387779e-16          0.272727   
cluster1          0.196078            2.156863e-01          0.882353   

          Australian Restaurant  Brazilian Restaurant  \
cluster0           1.040834e-17              0.006993   
cluster1           1.960784e-02              0.196078   

          Cajun / Creole Restaurant  Cantonese Restaurant  \
cluster0                   0.020979          5.551115e-17   
cluster1                   0.098039          1.372549e-01   

          Caribbean Restaurant  Chinese Restaurant  Colombian Restaurant  \
cluster0              0.566434            1.146853              0.013986   
cluster1              0.490196            2.607843              0.058824   

          Comfort Food Restaurant  Cuban Restaurant  Czech Restaurant  \
cluster0                 0.020979          0.013986      1.040834e-17   
cluster1                 0.019608          0.078431      1.960784e-02   

          Dim Sum Restaurant  Dumpling Restaurant  \
cluster0            0.013986             0.013986   
cluster1            0.117647             0.215686   

          Eastern European Restaurant  Egyptian Restaurant  \
cluster0                     0.041958         1.040834e-17   
cluster1                     0.156863         1.960784e-02   

          Empanada Restaurant  Falafel Restaurant  Fast Food Restaurant  \
cluster0             0.013986            0.006993              1.090909   
cluster1             0.098039            0.019608              1.215686   

          Filipino Restaurant  French Restaurant  German Restaurant  \
cluster0             0.027972           0.013986           0.034965   
cluster1             0.156863           0.176471           0.019608   

          Gluten-free Restaurant  Greek Restaurant  Halal Restaurant  \
cluster0            1.040834e-17          0.076923          0.013986   
cluster1            1.960784e-02          0.647059          0.117647   

          Himalayan Restaurant  Hotpot Restaurant  Indian Restaurant  \
cluster0          8.673617e-17       8.326673e-17           0.132867   
cluster1          1.176471e-01       1.568627e-01           1.215686   

          Indonesian Restaurant  Italian Restaurant  Japanese Restaurant  \
cluster0           4.163336e-17            1.188811             0.244755   
cluster1           7.843137e-02            2.549020             0.823529   

          Korean Restaurant  Kosher Restaurant  Latin American Restaurant  \
cluster0           0.027972       2.081668e-17                   0.258741   
cluster1           1.549020       3.921569e-02                   1.294118   

          Malay Restaurant  Mediterranean Restaurant  Mexican Restaurant  \
cluster0      4.163336e-17                  0.048951            0.475524   
cluster1      7.843137e-02                  0.235294            2.215686   

          Middle Eastern Restaurant  Modern Greek Restaurant  \
cluster0                   0.076923             1.040834e-17   
cluster1                   0.294118             1.960784e-02   

          New American Restaurant  Paella Restaurant  Pakistani Restaurant  \
cluster0                 0.034965           0.013986              0.020979   
cluster1                 0.137255           0.019608              0.039216   

          Persian Restaurant  Peruvian Restaurant  Polish Restaurant  \
cluster0        1.040834e-17             0.062937       1.040834e-17   
cluster1        1.960784e-02             0.490196       1.960784e-02   

          Portuguese Restaurant  Ramen Restaurant  Restaurant  \
cluster0           6.993007e-03      8.673617e-17    0.307692   
cluster1          -2.602085e-18      1.176471e-01    0.686275   

          Romanian 

### The Total and Total Sum of cluster 0 has smallest value. It shows that the market is not saturated.

***Dataframe with Neighborhood,Cluster No and Total Sum***

In [61]:
BQS_results_merged = pd.DataFrame(BQS_restaurant_grouped['Neighborhood'],)

BQS_results_merged['Total'] = BQS_restaurant_grouped['Total']
BQS_results_merged = BQS_results_merged.assign(Cluster_Labels = kmeans.labels_)

In [ ]:
print(BQS_results_merged.shape)
BQS_results_merged

***Merge BQS_results_merged with BQS_Geo***

In [63]:
BQS_merged = BQS_Geo

BQS_merged = BQS_merged.join(BQS_results_merged.set_index('Neighborhood'), on='Neighborhood')

print(BQS_merged.shape)
BQS_merged.head(10) # check the last columns!

(196, 6)


Borough    Neighborhood   Latitude  Longitude  Total  Cluster_Labels
0   Bronx       Wakefield  40.894705 -73.847201     14               0
1   Bronx      Co-op City  40.874294 -73.829939     13               0
2   Bronx     Eastchester  40.887556 -73.827806     12               0
3   Bronx       Fieldston  40.895437 -73.905643      9               0
4   Bronx       Riverdale  40.890834 -73.912585     10               0
5   Bronx     Kingsbridge  40.881687 -73.902818     20               1
6   Bronx        Woodlawn  40.898273 -73.867315      9               0
7   Bronx         Norwood  40.877224 -73.879391     20               1
8   Bronx  Williamsbridge  40.881039 -73.857446     14               0
9   Bronx      Baychester  40.866858 -73.835798     15               0

***Finally, let's visualize the resulting clusters***

In [ ]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i+x+(i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(BQS_merged['Latitude'], BQS_merged['Longitude'], BQS_merged['Neighborhood'], BQS_merged['Cluster_Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## List Neighborhoods of Interest in New York City - Bronx, Queens and Staten Island

## Cluster 1 : Saturated Markets

In [ ]:
BQS_merged[BQS_merged['Cluster_Labels'] == 1].reset_index(drop=True)

## Cluster 0 : Untapped Markets

In [66]:
BQS_merged[BQS_merged['Total'] == 0].reset_index(drop=True)

Borough  Neighborhood   Latitude  Longitude  Total  Cluster_Labels
0          Bronx  Clason Point  40.806551 -73.854144      0               0
1  Staten Island     Todt Hill  40.597069 -74.111329      0               0
2  Staten Island   South Beach  40.580247 -74.079553      0               0
3  Staten Island    Port Ivory  40.639683 -74.174645      0               0
4  Staten Island    Bloomfield  40.605779 -74.187256      0               0